watertap/core/plugins/tests/test_solvers.py

```diff
@@ -0,0 +1,148 @@
+    def m(self):
+        m = pyo.ConcreteModel()
+        m.a = pyo.Var(initialize=1)
+        m.b = b = pyo.Block()
+        b.a = pyo.Var([1,2])
+
+        m.c = pyo.Constraint(expr=(0, 1./(m.a**2) ,1))
+        b.c = pyo.Constraint([1,2], rule= lambda b,i: (i-1, b.a[i], i+2))
+        b.d = pyo.Constraint(expr=b.a[1]**4 + b.a[2]**4 <= 4)
+        set_scaling_factor(b.d, 1e6)
+
+        b.o = pyo.Expression(expr=sum(b.a)**2)
+
+        m.o = pyo.Objective(expr=m.a+b.o)
+
+        return m
+
+    @pytest.fixture(scope="class")
+    def s(self):
+        return pyo.SolverFactory('ipopt-watertap')
+    
+    @pytest.mark.unit
+    def test_pyomo_registration(self, s):
+        assert s.__class__ is IpoptWaterTAP
+
+    @pytest.mark.unit
+    def test_idaes_registration(self):
+        assert get_solver().__class__ is IpoptWaterTAP
+
+    @pytest.mark.unit
+    def test_presolve_scales_constraints(self, m, s):
+        s._presolve(m, tee=True)
+        for c, sf in s._scaling_cache:
+            if c is m.b.d:
+                assert sf == 1e6
+            else:
+                assert sf is None
+        
+        cons_with_sf = [c for c,_ in constraints_with_scale_factor_generator(m)]
+        assert m.c in cons_with_sf
+        assert m.b.c[1] in cons_with_sf
+        assert m.b.c[2] in cons_with_sf
+        assert m.b.d in cons_with_sf
+
+        assert s._model is m
+
+    @pytest.mark.unit
+    def test_postsolve_unscaled_constraints_cleanup(self, m, s):
+        assert hasattr(s, '_postsolve')
+        # for the Pyomo implementation
+        try:
+            s._postsolve()
+        except AttributeError:
+            pass
+
+        assert not hasattr(s, '_scaling_cache')
+        
+        cons_with_sf = list(constraints_with_scale_factor_generator(m))
+        assert cons_with_sf == [(m.b.d, 1e6)]
+
+        assert not hasattr(s, '_model')
+
+    @pytest.mark.unit
+    def test_option_absorption(self, m, s):
+        s.options['ignore_variable_scaling'] = True
+        results = s.solve(m, tee=True)
+        pyo.assert_optimal_termination(results)
+        del s.options['ignore_variable_scaling']
+
+    @pytest.mark.unit
+    def test_get_option(self, s):
+        s.options['ignore_constraint_scaling'] = True
+
+        assert s._get_option('ignore_constraint_scaling', False) is True
+        assert 'ignore_constraint_scaling' not in s.options
+        assert s._get_option('ignore_constraint_scaling', False) is False
+
+    @pytest.mark.unit
+    def test_presolve_passthrough(self, m, s):
+        s.options["nlp_scaling_method"] = "gradient-based"
+        s._presolve(m, tee=True)
+        assert not hasattr(s, "_scaling_cache")
+        s.options["nlp_scaling_method"] = "user-scaling"
+
+    @pytest.mark.unit
+    def test_passthrough_positive(self, m, s):
+        s.options["nlp_scaling_method"] = "gradient-based"
+        pyo.assert_optimal_termination(s.solve(m, tee=True))
+        del s.options["nlp_scaling_method"]
```

Comment:
Why delete the option here but in line 107 reassign to `user-scaling`?

Reply:
Oh, because in line 107, you don't want to delete after a presolve. In this test you're doing the full solve. Right?

Reply:
Probably safer to do a delete on line 107.